README.pod
=pod

=head1 NAME

Mojolicious::Plugin::Log::Any - Use other loggers in a Mojolicious application

=head1 SYNOPSIS

  package MyApp;
  use Mojo::Base 'Mojolicious';
  
  sub startup {
    my $self = shift;
    
    # Log::Any (default)
    use Log::Any::Adapter {category => 'MyApp', message_separator => ' '}, 'Syslog';
    $self->plugin('Log::Any');
    
    # Log::Contextual
    use Log::Contextual::WarnLogger;
    use Log::Contextual -logger => Log::Contextual::WarnLogger->new({env_prefix => 'MYAPP'});
    $self->plugin('Log::Any' => {logger => 'Log::Contextual});
    
    # Log::Dispatch
    use Log::Dispatch;
    my $logger = Log::Dispatch->new(outputs => ['File::Locked',
      min_level => 'warning',
      filename  => '/path/to/file.log',
      mode      => 'append',
      newline   => 1,
      callbacks => sub { my %p = @_; '[' . localtime() . '] ' . $p{message} },
    ]);
    $self->plugin('Log::Any' => {logger => $logger});
    
    # Log::Dispatchouli
    use Log::Dispatchouli;
    my $logger = Log::Dispatchouli->new({ident => 'MyApp', facility => 'daemon', to_file => 1});
    $self->plugin('Log::Any' => {logger => $logger});
    
    # Log::Log4perl
    use Log::Log4perl;
    Log::Log4perl->init($self->home->child('log.conf')->to_string);
    $self->plugin('Log::Any' => {logger => 'Log::Log4perl'});
  }
  
  # or in a Mojolicious::Lite app
  use Mojolicious::Lite;
  use Log::Any::Adapter {category => 'Mojolicious::Lite'}, File => app->home->child('myapp.log'), log_level => 'info';
  plugin 'Log::Any';

=head1 DESCRIPTION

L<Mojolicious::Plugin::Log::Any> is a L<Mojolicious> plugin that redirects the
application logger to pass its log messages to an external logging framework
using L<Mojo::Log::Role::AttachLogger/"attach_logger">. By default, L<Log::Any>
is used, but a different framework or object may be specified. For L<Log::Any>
or L<Log::Log4perl>, log messages are dispatched with a category of the
application class name, which is C<Mojolicious::Lite> for lite applications.

The default behavior of the L<Mojo::Log> object to filter messages by level,
keep history, prepend a timestamp, and write log messages to a file or STDERR
will be suppressed, by setting the application log level to C<debug> or
C<trace> (the lowest level) and removing the default L<Mojo::Log/"message">
handler. It is expected that the logging framework output handler will be
configured to handle these details as necessary. If you want to customize how
the logging framework is attached, use L<Mojo::Log::Role::AttachLogger>
directly.

=head1 METHODS

L<Mojolicious::Plugin::Log::Any> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Mojolicious->new);
  $plugin->register(Mojolicious->new, {logger => $logger});

Register logger in L<Mojolicious> application. Takes the following options:

=over

=item logger

Logging framework or object to pass log messages to, of a type recognized by
L<Mojo::Log::Role::AttachLogger/"attach_logger">. Defaults to C<Log::Any>.

=item category

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">. Defaults to
the application name.

=item prepend_level

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">.

=item message_separator

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">.

=back

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::Log>, L<Mojo::Log::Role::AttachLogger>

=cut


Changes
{{$NEXT}}

v2.0.0    2021-08-17 14:05:38 EDT
  - Support "trace" log level with Mojolicious 9.20 or newer.
    - An attached Log::Dispatch logger will log "trace" messages at the "debug" level.
    - An attached Log::Dispatchouli logger will log "trace" messages with "log_debug".
    - You may need to enable the "trace" level in other types of attached loggers.
    - Mojolicious::Plugin::Log::Any now sets the log level to "trace" if available.

v1.0.3    2021-06-24 19:29:25 EDT
  - Add message_separator option to attach_logger and plugin to allow changing
    the separator used to join messages (#5, robrwo)

v1.0.2    2021-06-16 00:49:01 EDT
  - Update IRC metadata to libera.chat

v1.0.1    2020-03-31 12:51:35 EDT
  - Avoid stringifying message when passed as a single reference with
    prepend_level false, such as when logging exception objects.

v1.0.0    2019-08-12 18:44:18 EDT
  - Add prepend_level option to attach_logger and plugin to allow disabling the
    prepending of the log level to messages (#2)

0.005     2018-12-05 00:33:32 EST
  - Fix tests to not rely on shorthand for applying role
  - Mojolicious 8.06 changed Mojo::Log to not send the 'message' event when a
    log message would be filtered by the log level, so the plugin now explicitly
    sets the application log level to 'debug' to preserve the old (documented)
    behavior of always forwarding to the attached logger

0.004     2018-02-06 23:16:16 EST
  - Require Log::Any 1.00 for Log::Any testing

0.003     2018-02-06 14:05:06 EST
  - Require Log::Contextual 0.008001 when attaching it as a logger
  - Import slog_ functions individually from Log::Contextual

0.002     2017-12-05 02:28:09 EST
  - Quick breaking change: Unsubscribe the default message handler in the
    plugin, not the role

0.001     2017-12-05 02:03:16 EST
  - First release


Build.PL
# This Build.PL for Mojolicious-Plugin-Log-Any was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojolicious-Plugin-Log-Any was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.017.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "Use other loggers in a Mojolicious application",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojolicious-Plugin-Log-Any",
      "dist_version" => "v1.0.2",
      "license" => "artistic_2",
      "module_name" => "Mojolicious::Plugin::Log::Any",
      "recommends" => {
        "Log::Any" => 0
      },
      "recursive_test_files" => 1,
      "requires" => {
        "Carp" => 0,
        "Import::Into" => 0,
        "Module::Runtime" => 0,
        "Mojolicious" => "7.40",
        "Role::Tiny" => "2.000001",
        "Scalar::Util" => 0,
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88",
        "Test::Needs" => 0
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88",
      "Test::Needs" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}


lib/Mojolicious/Plugin/Log/Any.pm
package Mojolicious::Plugin::Log::Any;

use Mojo::Base 'Mojolicious::Plugin';

our $VERSION = 'v2.0.1';

sub register {
  my ($self, $app, $conf) = @_;
  
  my $logger = delete $conf->{logger} // 'Log::Any';
  my %opt = %$conf;
  $opt{category} //= ref $app;

  my $lowest_level;
  { local $@;
    $lowest_level = eval { require Mojolicious; Mojolicious->VERSION('9.20'); 1 } ? 'trace' : 'debug';
  }
  $app->log->with_roles('Mojo::Log::Role::AttachLogger')->level($lowest_level)
    ->unsubscribe('message')->attach_logger($logger, \%opt);
}

1;

=head1 NAME

Mojolicious::Plugin::Log::Any - Use other loggers in a Mojolicious application

=head1 SYNOPSIS

  package MyApp;
  use Mojo::Base 'Mojolicious';
  
  sub startup {
    my $self = shift;
    
    # Log::Any (default)
    use Log::Any::Adapter {category => 'MyApp', message_separator => ' '}, 'Syslog';
    $self->plugin('Log::Any');
    
    # Log::Contextual
    use Log::Contextual::WarnLogger;
    use Log::Contextual -logger => Log::Contextual::WarnLogger->new({env_prefix => 'MYAPP'});
    $self->plugin('Log::Any' => {logger => 'Log::Contextual});
    
    # Log::Dispatch
    use Log::Dispatch;
    my $logger = Log::Dispatch->new(outputs => ['File::Locked',
      min_level => 'warning',
      filename  => '/path/to/file.log',
      mode      => 'append',
      newline   => 1,
      callbacks => sub { my %p = @_; '[' . localtime() . '] ' . $p{message} },
    ]);
    $self->plugin('Log::Any' => {logger => $logger});
    
    # Log::Dispatchouli
    use Log::Dispatchouli;
    my $logger = Log::Dispatchouli->new({ident => 'MyApp', facility => 'daemon', to_file => 1});
    $self->plugin('Log::Any' => {logger => $logger});
    
    # Log::Log4perl
    use Log::Log4perl;
    Log::Log4perl->init($self->home->child('log.conf')->to_string);
    $self->plugin('Log::Any' => {logger => 'Log::Log4perl'});
  }
  
  # or in a Mojolicious::Lite app
  use Mojolicious::Lite;
  use Log::Any::Adapter {category => 'Mojolicious::Lite'}, File => app->home->child('myapp.log'), log_level => 'info';
  plugin 'Log::Any';

=head1 DESCRIPTION

L<Mojolicious::Plugin::Log::Any> is a L<Mojolicious> plugin that redirects the
application logger to pass its log messages to an external logging framework
using L<Mojo::Log::Role::AttachLogger/"attach_logger">. By default, L<Log::Any>
is used, but a different framework or object may be specified. For L<Log::Any>
or L<Log::Log4perl>, log messages are dispatched with a category of the
application class name, which is C<Mojolicious::Lite> for lite applications.

The default behavior of the L<Mojo::Log> object to filter messages by level,
keep history, prepend a timestamp, and write log messages to a file or STDERR
will be suppressed, by setting the application log level to C<debug> or
C<trace> (the lowest level) and removing the default L<Mojo::Log/"message">
handler. It is expected that the logging framework output handler will be
configured to handle these details as necessary. If you want to customize how
the logging framework is attached, use L<Mojo::Log::Role::AttachLogger>
directly.

=head1 METHODS

L<Mojolicious::Plugin::Log::Any> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Mojolicious->new);
  $plugin->register(Mojolicious->new, {logger => $logger});

Register logger in L<Mojolicious> application. Takes the following options:

=over

=item logger

Logging framework or object to pass log messages to, of a type recognized by
L<Mojo::Log::Role::AttachLogger/"attach_logger">. Defaults to C<Log::Any>.

=item category

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">. Defaults to
the application name.

=item prepend_level

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">.

=item message_separator

Passed through to L<Mojo::Log::Role::AttachLogger/"attach_logger">.

=back

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::Log>, L<Mojo::Log::Role::AttachLogger>


lib/Mojo/Log/Role/AttachLogger.pm
package Mojo::Log::Role::AttachLogger;

use Role::Tiny;
use Carp ();
use Import::Into ();
use Module::Runtime ();
use Scalar::Util ();

our $VERSION = 'v2.0.1';

our @CARP_NOT = 'Mojolicious::Plugin::Log::Any';

requires 'on';

sub attach_logger {
  my ($self, $logger, $opt) = @_;
  Carp::croak 'No logger passed' unless defined $logger;
  my ($category, $prepend, $separator);
  if (ref $opt) {
    ($category, $prepend, $separator) = @$opt{qw[ category prepend_level message_separator ]};
  } else {
    $category = $opt;
  }
  $category //= 'Mojo::Log';
  $prepend //= 1;
  $separator //= "\n";

  my $do_log;
  if (Scalar::Util::blessed($logger)) {
    if ($logger->isa('Log::Any::Proxy')) {
      $do_log = sub {
        my ($self, $level, @msg) = @_;
        my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
        $msg = "[$level] $msg" if $prepend;
        $logger->$level($msg);
      };
    } elsif ($logger->isa('Log::Dispatch')) {
      $do_log = sub {
        my ($self, $level, @msg) = @_;
        my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
        $msg = "[$level] $msg" if $prepend;
        $level = 'critical' if $level eq 'fatal';
        $level = 'debug' if $level eq 'trace';
        $logger->log(level => $level, message => $msg);
      };
    } elsif ($logger->isa('Log::Dispatchouli') or $logger->isa('Log::Dispatchouli::Proxy')) {
      $do_log = sub {
        my ($self, $level, @msg) = @_;
        my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
        $msg = "[$level] $msg" if $prepend;
        return $logger->log_debug($msg) if $level eq 'trace' or $level eq 'debug';
        # hacky but we don't want to use log_fatal because it throws an
        # exception, we want to allow real exceptions to propagate, and we
        # can't localize a call to set_muted
        local $logger->{muted} = 0 if $level eq 'fatal' and $logger->get_muted;
        $logger->log($msg);
      };
    } elsif ($logger->isa('Mojo::Log')) {
      $do_log = sub {
        my ($self, $level, @msg) = @_;
        $logger->$level(@msg);
      };
    } else {
      Carp::croak "Unsupported logger object class " . ref($logger);
    }
  } elsif ($logger eq 'Log::Any') {
    require Log::Any;
    $logger = Log::Any->get_logger(category => $category);
    $do_log = sub {
      my ($self, $level, @msg) = @_;
      my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
      $msg = "[$level] $msg" if $prepend;
      $logger->$level($msg);
    };
  } elsif ($logger eq 'Log::Log4perl') {
    require Log::Log4perl;
    $logger = Log::Log4perl->get_logger($category);
    $do_log = sub {
      my ($self, $level, @msg) = @_;
      my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
      $msg = "[$level] $msg" if $prepend;
      $logger->$level($msg);
    };
  } elsif ($logger eq 'Log::Contextual' or "$logger"->isa('Log::Contextual')) {
    Module::Runtime::require_module("$logger");
    Log::Contextual->VERSION('0.008001');
    my %functions = map { ($_ => "slog_$_") } qw(trace debug info warn error fatal);
    "$logger"->import::into(ref($self), values %functions);
    $do_log = sub {
      my ($self, $level, @msg) = @_;
      my $msg = @msg > 1 ? join($separator, @msg) : $msg[0];
      $msg = "[$level] $msg" if $prepend;
      $self->can($functions{$level})->($msg);
    };
  } else {
    Carp::croak "Unsupported logger class $logger";
  }
  
  $self->on(message => $do_log);
  
  return $self;
}

1;

=head1 NAME

Mojo::Log::Role::AttachLogger - Use other loggers for Mojo::Log

=head1 SYNOPSIS

  use Mojo::Log;
  my $log = Mojo::Log->with_roles('+AttachLogger')->new->unsubscribe('message');
  
  # Log::Any
  use Log::Any::Adapter {category => 'Mojo::Log', message_separator => ' '}, 'Syslog';
  $log->attach_logger('Log::Any', 'Some::Category');
  
  # Log::Contextual
  use Log::Contextual::WarnLogger;
  use Log::Contextual -logger => Log::Contextual::WarnLogger->new({env_prefix => 'MYAPP'});
  $log->attach_logger('Log::Contextual');
  
  # Log::Dispatch
  use Log::Dispatch;
  my $logger = Log::Dispatch->new(outputs => ['File::Locked',
    min_level => 'warning',
    filename  => '/path/to/file.log',
    mode      => 'append',
    newline   => 1,
    callbacks => sub { my %p = @_; '[' . localtime() . '] ' . $p{message} },
  ]);
  $log->attach_logger($logger);
  
  # Log::Dispatchouli
  use Log::Dispatchouli;
  my $logger = Log::Dispatchouli->new({ident => 'MyApp', facility => 'daemon', to_file => 1});
  $log->attach_logger($logger);
  
  # Log::Log4perl
  use Log::Log4perl;
  Log::Log4perl->init('/path/to/log.conf');
  $log->attach_logger('Log::Log4perl', 'Some::Category');
  
=head1 DESCRIPTION

L<Mojo::Log::Role::AttachLogger> is a L<Role::Tiny> role for L<Mojo::Log> that
redirects log messages to an external logging framework. L</"attach_logger">
currently recognizes the strings C<Log::Any>, C<Log::Contextual>,
C<Log::Log4perl>, and objects of the classes C<Log::Any::Proxy>,
C<Log::Dispatch>, C<Log::Dispatchouli>, and C<Mojo::Log>.

The default L<Mojo::Log/"message"> event handler is not suppressed by
L</"attach_logger">, so if you want to suppress the default behavior, you
should unsubscribe from the message event first. Unsubscribing from the message
event will also remove any loggers attached by L</"attach_logger">.

Since L<Mojolicious> 8.06, the L<Mojo::Log/"message"> event will not be sent
for messages below the log level set in the L<Mojo::Log> object, so the
attached logger will only receive log messages exceeding the configured level.

Since L<Mojolicious> 9.20, the C<trace> log level is supported though it may be
mapped to C<debug> on some loggers.

L<Mojolicious::Plugin::Log::Any> can be used to attach a logger to the
L<Mojolicious> application logger and suppress the default message event
handler.

=head1 METHODS

L<Mojo::Log::Role::AttachLogger> composes the following methods.

=head2 attach_logger

  $log = $log->attach_logger($logger, $options);

Subscribes to L<Mojo::Log/"message"> and passes log messages to the given
logging framework or object. The second argument is optionally a category
(default C<Mojo::Log>) or hashref of options. The log level will be prepended
to the message in square brackets (except when passing to another L<Mojo::Log>
object, or L</"prepend_level"> is false).

The following loggers are recognized:

=over

=item Log::Any

The string C<Log::Any> will use a global L<Log::Any> logger with the specified
category (defaults to C<Mojo::Log>).

=item Log::Any::Proxy

A L<Log::Any::Proxy> object can be passed directly and will be used for logging
in the standard manner, using the object's existing category.

=item Log::Contextual

The string C<Log::Contextual> will use the global L<Log::Contextual> logger.
Package loggers are not supported. Note that L<Log::Contextual/"with_logger">
may be difficult to use with L<Mojolicious> logging due to the asynchronous
nature of the dispatch cycle.

=item Log::Dispatch

A L<Log::Dispatch> object can be passed to be used for logging. The C<fatal>
log level will be mapped to C<critical>, and the C<trace> and C<debug> log
levels will both be mapped to C<debug>.

=item Log::Dispatchouli

A L<Log::Dispatchouli> object can be passed to be used for logging. The
C<fatal> log level will log messages even if the object is C<muted>, but an
exception will not be thrown as L<Log::Dispatchouli/"log_fatal"> normally does.
The C<trace> and C<debug> log levels will be logged with
L<Log::Dispatchouli/"log_debug">.

=item Log::Log4perl

The string C<Log::Log4perl> will use a global L<Log::Log4perl> logger with the
specified category (defaults to C<Mojo::Log>).

=item Mojo::Log

Another L<Mojo::Log> object can be passed to be used for logging.

=back

The following options are supported:

=over

=item category

Category name (defaults to Mojo::Log).

=item prepend_level

Prepend the log level to messages in the form C<[$level]> (default for
non-L<Mojo::Log> loggers). Set false to disable.

=item message_separator

String to separate multiple messages. Defaults to newline.

=back

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::Log>, L<Log::Any>, L<Log::Contextual>, L<Log::Dispatch>,
L<Log::Dispatchouli>, L<Log::Log4perl>
